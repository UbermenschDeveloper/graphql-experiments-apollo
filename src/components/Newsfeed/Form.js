import React, {PureComponent} from 'react'
import PropTypes from 'prop-types'
import injectSheet from '../../utils/jss'
import Card from '../ui/Card'
import Loader from '../ui/Loader'
import {ButtonPrimary} from '../ui/Button'
import {Input, Textarea} from '../ui/Input'
import {H2} from '../typography'
import {getErrorByField} from '../../utils/validator'

const styles = {
  container: {
    padding: 40,
    marignBottom: 20,
  },
  row: {
    marginBottom: 20,
  },
  actions: {
    display: 'flex',
    justifyContent: 'flex-end',
    marginTop: 20,
  },
}

class NewsfeedForm extends PureComponent {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    title: PropTypes.string,
    content: PropTypes.string,
    errors: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.string)),
    onSubmit: PropTypes.func.isRequired,
    sending: PropTypes.bool,
  }

  static defaultProps = {
    title: '',
    content: '',
    sending: false,
  }

  constructor(props) {
    super(props)

    this.state = {
      title: props.title,
      content: props.content,
    }
  }

  componentWillReceiveProps(nextProps) {
    // In case if form was send - reset all fields
    if (this.props.sending && !nextProps.sending) {
      this.setState({
        title: '',
        content: '',
      })
    }
  }

  handleTitleInput(value) {
    this.setState({title: value})
  }

  handleContentInput(value) {
    this.setState({content: value})
  }

  handleSubmit() {
    const {title, content} = this.state

    this.props.onSubmit({
      title,
      content,
    })
  }

  render() {
    const {classes, sending, errors} = this.props
    const {title, content} = this.state

    return (
      <Loader active={sending}>
        <Card className={classes.container}>
          <H2>What's on your mind?</H2>
          {/*
          <div className={classes.row}>
            <Input
              label="Title"
              value={title}
              error={getErrorByField(errors, 'title')}
              onInput={(value) => {
                this.handleTitleInput(value)
              }}
            />
          </div>
          */}
          <div className={classes.row}>
            <Textarea
              placeholder="Your post is very important for us!"
              value={content}
              error={getErrorByField(errors, 'content')}
              onInput={(value) => {
                this.handleContentInput(value)
              }}
            />
          </div>
          <div className={classes.actions}>
            <ButtonPrimary
              onClick={() => {
                this.handleSubmit()
              }}
            >
              Send
            </ButtonPrimary>
          </div>
        </Card>
      </Loader>
    )
  }
}

export default injectSheet(styles)(NewsfeedForm)
