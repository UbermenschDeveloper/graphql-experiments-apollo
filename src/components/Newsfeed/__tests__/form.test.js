import React from 'react'
import {spy} from 'sinon'
import {
  renderWithJss as render,
  mountWithJss as mount,
  extractFromJssWrapper,
} from '../../../utils/testSuite'
import Form from '../Form'

describe('FeedbackForm', () => {
  it('Should render correctly', () => {
    // Call InnerComponent for making snapshot of real component
    // instead of JSS-wrapped one
    const form = render(
      <Form.InnerComponent classes={{}} onSubmit={() => {}} />,
    )
    expect(form).toMatchSnapshot()
  })

  it('Should pass data via callback on submit click', () => {
    const onSubmit = spy()
    const form = mount(<Form onSubmit={onSubmit} />)

    form.find('Button').simulate('click')
    expect(onSubmit.called).toBeTruthy()
  })

  /**
   * Don't know why, but "form.setState"
   * doesn't update state in real world, and in onSubmit callback
   * arrives old data.
   */
  it('Should pass field values on submit', () => {
    let data = {}

    const form = mount(
      <Form
        onSubmit={(values) => {
          data = values
        }}
      />,
    )
    const button = form.find('.actions').children()

    button.simulate('click')
    expect(data).toEqual({
      title: '',
      content: '',
    })

    // Update state with new data
    extractFromJssWrapper(form)
      .instance()
      .setState({
        title: 'Title',
        content: 'Content',
      })

    button.simulate('click')

    expect(data).toEqual({
      title: 'Title',
      content: 'Content',
    })
  })
})
