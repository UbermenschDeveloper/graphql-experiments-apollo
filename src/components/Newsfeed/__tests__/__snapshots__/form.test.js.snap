// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm Should render correctly 1`] = `
<div
  class="loader"
>
  <div
    class="card"
  >
    <h2
      class="heading"
    >
      Leave your feedback
    </h2>
    <p
      class="paragraph"
    >
      Your feedback is very important for us!
    </p>
    <div>
      <div
        class="default"
      >
        <div
          class="label font"
        >
          Title
        </div>
        <div
          class="container"
        >
          <input
            class="value font"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="default"
      >
        <div
          class="label font"
        >
          Body
        </div>
        <div
          class="container"
        >
          <textarea
            class="value font"
            rows="1"
          />
        </div>
      </div>
    </div>
    <div>
      <button
        class="button primary button"
      >
        <span
          class="wrap"
        >
          <span
            class="text relative"
          >
            Send
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
